### PR TITLE
Add new Queue and Stack APIs

### LINKED ISSUE #4316: Add TryDequeue to Queue
[`System.Collections.Generic.Queue`](https://github.com/dotnet/corefx/blob/master/src/System.Collections/src/System/Collections/Generic/Queue.cs) does not have a `TryDequeue` method, however its `Dequeue` is a throwing method so the following pattern always has to be used:

```csharp
T item ...;
if (queue.Count > 0)
{
    item = queue.Dequeue();
    ...
}
```
with extra boilerplate for for whatever no item handling; this would be better as:
```csharp
T item;
if (queue.TryDequeue(out item)
{
    ...
}
```
As per `ConcurrentQueue` or `TryGet` of Dictionary

### PR DESCRIPTION
**Pull request by hughbe**
Fixes #4316

Also, adding APIs to .NET core was something of a soul destroying experience for me - very confusing as there are so many things to change and define and add. #11689 was helpful once I discovered it, but I ended up changing things without really understanding what the impact was.

/cc @benadams @Page-Not-Found @weshaggard @stephentoub @ianhays

### CODE REVIEW

**Comment by benaadams** [comment]
@hughbe Thank you 👍

**Comment by hughbe** [comment]
I'm getting test failures of `System.IO.FileNotFoundException : Could not load file or assembly 'System.Runtime.Serialization.Formatters, Version=4.0.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.` 
Not sure if I've broken this

**Comment by ianhays** [comment]
@joperezr

**Comment by ianhays** [comment]
>I'm getting test failures of System.IO.FileNotFoundException : Could not load file or assembly 'System.Runtime.Serialization.Formatters, Version=4.0.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified. 
Not sure if I've broken this or the CI is acting up - tests pass locally

I ran into this issue as well in my pr for FSW API additions: https://github.com/dotnet/corefx/pull/11785/files#diff-fd8e915288e5326816b99f27c89120deR31

I just added the library to the project.json for netstandard1.7.

**Comment by hughbe** [comment]
Ah ok - I'll try that. Seems a bit weird that it suddenly breaks though when adding a new target framework

**Comment by hughbe** [comment]
@ianhays It seems that this is already added to the project.. https://github.com/hughbe/corefx/blob/c815b7677764864a57a4cae3939fd6cf56ac5e04/src/System.Collections/tests/project.json#L28

**Comment by stephentoub** [comment]
The implementations of the new members LGTM.

**Comment by hughbe** [comment]
Thanks. @ianhays i ran the tests locally (msbuild /T:BuildAndTest /P:TargetGroup=netstandard1.7) and I got these errors), so I can reproduce the CI errors when specifying the target group. 
Unfortunately I don't understand the error message as the library is already referenced!

**Comment by ianhays** [comment]
>Also, adding APIs to .NET core was something of a soul destroying experience for me - very confusing as there are so many things to change and define and add.

Sorry about your trouble with the API addition Hugh. It's a pretty complicated process right now with the update to netstandard1.7, particularly since some libraries have been updated while others are still building netstandard1.3. It's still very much in flux, so I don't think we have any docs for it (at least I'm not aware of any).

System.Collections is one that has already been updated to netstandard1.7, so you shouldn't need to do any of the structural changes around versioning from this PR. 

> #11689 was helpful once I discovered it, but I ended up changing things without really understanding what the impact was.

Unfortunately https://github.com/dotnet/corefx/pull/11689 is a special case where Jeremy wanted some components to be exposed only for .NET Core (netcoreapp1.1) and not UAP (uap10.1) or net463. For most cases when adding API, we want to add those for all platforms that are adhere to netstandard1.7 (the previously mentioned), not just a particular one. For System.Collections specifically, there aren't any components that are runtime specific so the src build should be built for netstandard1.7. If you have any more in-depth questions then @joperezr can probably explain the TargetGroup/TestTFM structure better than I can since he's the one who explained it to me yesterday.

>nevermind, i see now that you already have that. I'll try to repro locally.

I pulled your changes, removed the netcoreapp1.1 structural stuff, and pushed a [working copy to my repo](https://github.com/ianhays/corefx/commit/dca90fa27dd98df52b06f2662a71135a334d3a63) that you can use as an example of what to do when a project has already been updated to build for netstandard1.7. I'd recommend you build the test builds file (```msbuild System.Collections.Tests.builds```) without args to test your changes, as that is how the CI will be testing them and it ensures you cover all of the configs.

File: src/System.Collections/tests/project.json [hunk 1]
```diff
@@ -27,7 +27,9 @@
       "dependencies": {
         "System.Runtime.Serialization.Formatters": "4.3.0-beta-24522-03"
       }
-    }
+    },
+    "netcoreapp1.1": {},
+    "netcoreapp1.0": {}
```
Comment (weshaggard) [inline src/System.Collections/tests/project.json]: Why did you need to add netcoreapp1.0?

File: src/System.Collections/tests/Generic/Queue/Queue.Generic.Tests.cs [hunk 2]
```diff
@@ -149,6 +149,30 @@ public void Queue_Generic_EnqueueAndDequeue(int capacity, int items)
             Assert.Equal(itemToAdd, q.Dequeue());
         }
 
+#if netcoreapp11
```
Comment (weshaggard) [inline src/System.Collections/tests/Generic/Queue/Queue.Generic.Tests.cs]: My preference would be to have these netcoreapp11 specific tests in their own file instead of using #ifdefs.

**Comment by weshaggard** [comment]
cc @ericstj

File: src/System.Collections/tests/project.json (hunk 1 above)
Comment (hughbe) [inline src/System.Collections/tests/project.json]: I saw it added in #11689 so replicated it

File: src/System.Collections/tests/System.Collections.Tests.csproj [hunk 3]
```diff
@@ -10,6 +10,7 @@
     <RootNamespace>System.Collections.Tests</RootNamespace>
     <DefineConstants Condition="'$(TargetGroup)'=='netstandard1.7'">$(DefineConstants);netstandard17</DefineConstants>
```
Comment (weshaggard) [inline src/System.Collections/tests/System.Collections.Tests.csproj]: You are going to need to add a build configuration in the System.Collections.Tests.builds file in order to get this netcoreapp1.1 build setup.

**Comment by hughbe** [comment]
@ianhays Thanks!
Unfortunately, I think I need to add these new APIs to netcoreapp, not netstandard1.7, as this API is currently .NET Core only?

File: src/System.Collections/tests/System.Collections.Tests.csproj (hunk 3 above)
Comment (weshaggard) [inline src/System.Collections/tests/System.Collections.Tests.csproj]: Something along the lines of:
```
    <Project Include="System.Collections.Tests.csproj">
      <TargetGroup>netcoreapp1.1</TargetGroup>
      <TestTFMs>netcoreapp1.1</TestTFMs>
    </Project>
```

Then to build and run the tests you either needs to `msbuild System.Collections.Tests.csproj /p:TargetGroup=netcoreapp1.1`

File: src/System.Collections/tests/project.json (hunk 1 above)
Comment (weshaggard) [inline src/System.Collections/tests/project.json]: You can remove the netcoreapp1.0 framework as you aren't building this project for that target framework.

Once you add the build configuration I suggested you should be able to test the build and run the tests for netcoreapp1.1. I suspect you will need to add:
```
 "System.Runtime.Serialization.Formatters": "4.3.0-beta-24522-03"
```
Under your netcoreapp1.1 section here as well, in order to get that dependency passed for the netcoreapp1.1 build.

**Comment by weshaggard** [comment]
>Unfortunately, I think I need to add these new APIs to netcoreapp, not netstandard1.7, as this API is currently .NET Core only?

**Comment by hughbe** [comment]
Okay! It all works guys, thanks for your help!

**Comment by hughbe** [comment]
Now the CentOs build must work too:
Test Innerloop CentOS7.1 Debug Build and Test
Test Innerloop CentOS7.1 Release Build and Test

File: src/System.Collections/tests/System.Collections.Tests.csproj [hunk 4]
```diff
@@ -10,6 +10,7 @@
...
     <NugetTargetMoniker Condition="'$(NugetTargetMoniker)'==''">.NETStandard,Version=v1.3</NugetTargetMoniker>
+    <DefineConstants Condition="'$(TargetGroup)' == 'netcoreapp1.1'">$(DefineConstants);netcoreapp11</DefineConstants>
```
Comment (weshaggard) [inline src/System.Collections/tests/System.Collections.Tests.csproj]: This define is no longer needed.

Comment (hughbe) [inline src/System.Collections/tests/System.Collections.Tests.csproj]: Good spot, thanks

**Comment by weshaggard** [comment]
Thanks for the contribution @hughbe.
